arch/xtensa/src/esp32s2/esp32s2_psram.c

```diff
@@ -0,0 +1,729 @@
+                        SPI_MEM_CS_HOLD_M | SPI_MEM_CS_SETUP_M);
+    }
+}
+
+static void psram_set_spi0_cache_cs_timing(psram_clk_mode_t clk_mode)
+{
+  if (clk_mode == PSRAM_CLK_MODE_NORM)
+    {
+      /* SPI0 SRAM Cache port */
+
+      SET_PERI_REG_BITS(SPI_MEM_SPI_SMEM_AC_REG(_SPI_CACHE_PORT),
+                        SPI_MEM_SPI_SMEM_CS_HOLD_TIME_V, 1,
+                        SPI_MEM_SPI_SMEM_CS_HOLD_TIME_S);
+      SET_PERI_REG_BITS(SPI_MEM_SPI_SMEM_AC_REG(_SPI_CACHE_PORT),
+                        SPI_MEM_SPI_SMEM_CS_SETUP_TIME_V, 0,
+                        SPI_MEM_SPI_SMEM_CS_SETUP_TIME_S);
+      SET_PERI_REG_MASK(SPI_MEM_SPI_SMEM_AC_REG(_SPI_CACHE_PORT),
+                        SPI_MEM_SPI_SMEM_CS_HOLD_M |
+                        SPI_MEM_SPI_SMEM_CS_SETUP_M);
+
+      /* SPI0 Flash Cache port */
+
+      SET_PERI_REG_BITS(SPI_MEM_CTRL2_REG(_SPI_CACHE_PORT),
+                        SPI_MEM_CS_HOLD_TIME_V, 0, SPI_MEM_CS_HOLD_TIME_S);
+
+      SET_PERI_REG_BITS(SPI_MEM_CTRL2_REG(_SPI_CACHE_PORT),
+                        SPI_MEM_CS_SETUP_TIME_V, 0,
+                        SPI_MEM_CS_SETUP_TIME_S);
+
+      SET_PERI_REG_MASK(SPI_MEM_USER_REG(_SPI_CACHE_PORT),
+                        SPI_MEM_CS_HOLD_M | SPI_MEM_CS_SETUP_M);
+    }
+  else
+    {
+      CLEAR_PERI_REG_MASK(SPI_MEM_USER_REG(_SPI_CACHE_PORT),
+                          SPI_CS_HOLD_M | SPI_CS_SETUP_M);
+    }
+}
+
+/* psram gpio init, different working frequency we have different
+ * solutions
+ */
+
+static void IRAM_ATTR psram_gpio_config(void)
+{
+  psram_io_t psram_io = PSRAM_IO_CONF_DEFAULT();
+  const uint32_t spiconfig = esp_rom_efuse_get_flash_gpio_info();
+
+  if (spiconfig == ESP_ROM_EFUSE_FLASH_DEFAULT_SPI)
+    {
+      /* FLASH pins(except wp / hd) are all configured via IO_MUX in rom. */
+    }
+  else
+    {
+      /* FLASH pins are all configured via GPIO matrix in ROM. */
+
+      psram_io.flash_clk_io       = EFUSE_SPICONFIG_RET_SPICLK(spiconfig);
+      psram_io.flash_cs_io        = EFUSE_SPICONFIG_RET_SPICS0(spiconfig);
+      psram_io.psram_spiq_sd0_io  = EFUSE_SPICONFIG_RET_SPIQ(spiconfig);
+      psram_io.psram_spid_sd1_io  = EFUSE_SPICONFIG_RET_SPID(spiconfig);
+      psram_io.psram_spihd_sd2_io = EFUSE_SPICONFIG_RET_SPIHD(spiconfig);
+      psram_io.psram_spiwp_sd3_io = esp_rom_efuse_get_flash_wp_gpio();
+    }
+
+  esp_rom_spiflash_select_qio_pins(psram_io.psram_spiwp_sd3_io, spiconfig);
+  g_psram_cs_io = psram_io.psram_cs_io;
+}
+
+int psram_get_size(void)
+{
+  int psram_size = PSRAM_SIZE_ID(g_psram_id);
+
+  if (PSRAM_IS_64MBIT_TRIAL(g_psram_id))
+    {
+      return PSRAM_SIZE_8MB;
+    }
+
+  switch (psram_size)
+    {
+      case PSRAM_EID_SIZE_64MBITS:
+        return PSRAM_SIZE_8MB;
+        break;
+
+      case PSRAM_EID_SIZE_32MBITS:
+        return PSRAM_SIZE_4MB;
+        break;
+
+      case PSRAM_EID_SIZE_16MBITS:
+        return PSRAM_SIZE_2MB;
+        break;
+
+      default:
+        return PSRAM_SIZE_16MB;
+        break;
```

Comment:
```suggestion
      case PSRAM_EID_SIZE_64MBITS:
        return PSRAM_SIZE_8MB;

      case PSRAM_EID_SIZE_32MBITS:
        return PSRAM_SIZE_4MB;

      case PSRAM_EID_SIZE_16MBITS:
        return PSRAM_SIZE_2MB;

      default:
        return PSRAM_SIZE_16MB;
```
`break` statement is unreachable.